### ISSUE TITLE
kIASKAppSettingChanged object in the notification

### ISSUE THREAD

**Issue by JanC** [futuretap/InAppSettingsKit#175]
Hi,
I see that the when the kIASKAppSettingChanged is posted, the object of the notification is the key of the value that changed. This key as again present in the userInfo:

`````
[[NSNotificationCenter defaultCenter] postNotificationName:kIASKAppSettingChanged
                                                        object:[toggle key]
                                                      userInfo:[NSDictionary dictionaryWithObject:[self.settingsStore        objectForKey:[toggle key]]
                                                                                           forKey:[toggle key]]];
`````
I think that passing the IASKAppSettingsController (self) as the object would be more appropriate. In the current implementation, the object handeling the notification has to have a reference to the IASKAppSettingsController. This goes imo agains the notification paradigm.

`````
pragma mark kIASKAppSettingChanged notification
- (void)settingDidChange:(NSNotification*)notification {
	if ([notification.object isEqual:@"AutoConnect"]) {
		IASKAppSettingsViewController *activeController = self.tabBarController.selectedIndex ? self.tabAppSettingsViewController : self.appSettingsViewController;
		BOOL enabled = (BOOL)[[notification.userInfo objectForKey:@"AutoConnect"] intValue];
		[activeController setHiddenKeys:enabled ? nil : [NSSet setWithObjects:@"AutoConnectLogin", @"AutoConnectPassword", nil] animated:YES];
	}
}
`````

If "self" is passed to the notification:
`````
[[NSNotificationCenter defaultCenter] postNotificationName:kIASKAppSettingChanged
                                                        object: elf
                                                      userInfo:[NSDictionary dictionaryWithObject:[self.settingsStore objectForKey:[toggle key]]
                                                                                           forKey:[toggle key]]];
`````
 the handeling would look just like that:

`````
#pragma mark kIASKAppSettingChanged notification
- (void)settingDidChange:(NSNotification*)notification {
    IASKAppSettingsViewController * activeController = notification.object; 
	if ([notification.userInfo objectForKey:@"AutoConnect"]) {
		BOOL enabled = (BOOL)[[notification.userInfo objectForKey:@"AutoConnect"] intValue];
		[activeController setHiddenKeys:enabled ? nil : [NSSet setWithObjects:@"AutoConnectLogin", @"AutoConnectPassword", nil] animated:YES];
	}
}
`````
No need to have a ivar in the object that is handeling the notification.

**Comment by GarethShapiro** [comment]
+1 for this.  

Is there as reason that self is not set as the object when kIASKAppSettingChanged is posted?

Locking the kIASKAppSettingChanged notifications down to specific instances of IASKAppSettingsViewController does avoid a common criticism of NSNotificationCenter, in that is has application scope and you can create unexpected side effects.

**Comment by futuretap** [comment]
You're right, it should send the instance of the view controller instead. Gladly accepting a PR!

**Comment by GarethShapiro** [comment]
I have a branch ready for you, if you are able to authorise me?

**Comment by futuretap** [comment]
Just create a pull request. Then I can merge it.

**Comment by GarethShapiro** [comment]
I can't push the branch.  Is it something on my side?  I've been having auth probs since going to 2-factor for work ...

**Comment by futuretap** [comment]
Just create a fork of IASK on GitHub, then push your changes to your branch. Then GitHub lets you create a  pull request for your branch.

**Comment by futuretap** [comment]
Thanks, merged.

**Comment by AndyWeinstein** [comment]
I think this may have been implemented inconsistently.
It also broke code that I inherited which tracks settings changes that can dynamically effect what settings should be available. That code was expected the object to be the setting that was changed. I can fix this issue in my code but I have to wonder if I was the only one affected and who spent a long time trying to figure this out.

I am new here so I'm treading carefully...
in IASKAppSettingsViewController there are four postNotification calls:
toggledValue, sliderChangedValue, _textChanged all have "self" for the object parameter
but textViewDidChange  has:
[[NSNotificationCenter defaultCenter] postNotificationName:kIASKAppSettingChanged
														object:textView.key
													  userInfo:@{textView.key: textView.text}];
	
There is also one in the IASKMultipleValueSelection which uses "self" for object.
